src/ImageSharp/Formats/Pbm/PlainEncoder.cs

```diff
@@ -0,0 +1,239 @@
+using SixLabors.ImageSharp.PixelFormats;
+
+namespace SixLabors.ImageSharp.Formats.Pbm
+{
+    /// <summary>
+    /// Pixel encoding methods for the PBM plain encoding.
+    /// </summary>
+    internal class PlainEncoder
+    {
+        private const int MaxLineLength = 70;
+        private const byte NewLine = 0x0a;
+        private const byte Space = 0x20;
+        private const byte Zero = 0x30;
+        private const byte One = 0x31;
+
+        private const int MaxCharsPerPixelBlackAndWhite = 2;
+        private const int MaxCharsPerPixelGrayscale = 4;
+        private const int MaxCharsPerPixelGrayscaleWide = 6;
+        private const int MaxCharsPerPixelRgb = 4 * 3;
+        private const int MaxCharsPerPixelRgbWide = 6 * 3;
+
+        private static readonly StandardFormat DecimalFormat = StandardFormat.Parse("D");
+
+        /// <summary>
+        /// Decode pixels into the PBM plain encoding.
+        /// </summary>
+        /// <typeparam name="TPixel">The type of input pixel.</typeparam>
+        /// <param name="configuration">The configuration.</param>
+        /// <param name="stream">The bytestream to write to.</param>
+        /// <param name="image">The input image.</param>
+        /// <param name="colorType">The ColorType to use.</param>
+        /// <param name="maxPixelValue">The maximum expected pixel value</param>
+        public static void WritePixels<TPixel>(Configuration configuration, Stream stream, ImageFrame<TPixel> image, PbmColorType colorType, int maxPixelValue)
+            where TPixel : unmanaged, IPixel<TPixel>
+        {
+            if (colorType == PbmColorType.Grayscale)
+            {
+                if (maxPixelValue < 256)
+                {
+                    WriteGrayscale(configuration, stream, image);
+                }
+                else
+                {
+                    WriteWideGrayscale(configuration, stream, image);
+                }
+            }
+            else if (colorType == PbmColorType.Rgb)
+            {
+                if (maxPixelValue < 256)
+                {
+                    WriteRgb(configuration, stream, image);
+                }
+                else
+                {
+                    WriteWideRgb(configuration, stream, image);
+                }
+            }
+            else
+            {
+                WriteBlackAndWhite(configuration, stream, image);
+            }
+        }
+
+        private static void WriteGrayscale<TPixel>(Configuration configuration, Stream stream, ImageFrame<TPixel> image)
+            where TPixel : unmanaged, IPixel<TPixel>
+        {
+            int width = image.Width;
+            int height = image.Height;
+            MemoryAllocator allocator = configuration.MemoryAllocator;
+            using IMemoryOwner<L8> row = allocator.Allocate<L8>(width);
+            Span<L8> rowSpan = row.GetSpan();
+            Span<byte> plainSpan = stackalloc byte[width * MaxCharsPerPixelGrayscale];
```

Comment:
We shouldn't be using `stackalloc` for larger numbers like this here and elsewhere. You're actually allocating more memory here than you use the `MemoryAllocator` to allocate the `L8` span. Might work in some environments but stackoverflow in others.

Here's a great blog post on `stackalloc` usage. https://vcsjones.dev/

Reply:
Interesting read. Reverted back to Allocator again.